src/software/power/geneva.cpp

```diff
@@ -1,32 +1,128 @@
 #include "geneva.h"
 
-hw_timer_t* Geneva::timer = nullptr;
+hw_timer_t* Geneva::timer                  = nullptr;
+volatile int Geneva::dir                   = 0;
+volatile int Geneva::count_a               = 0;
+volatile int Geneva::count_b               = 0;
+volatile int Geneva::prev_count_a          = 0;
+volatile int Geneva::prev_count_b          = 0;
+TbotsProto_Geneva_Slot Geneva::currentSlot = TbotsProto_Geneva_Slot_CENTRE;
+void (*volatile IRAM_ATTR Geneva::rotation_done_callback)() = NULL;
 
 Geneva::Geneva()
 {
-    timer = timerBegin(0, 80, true);
+    pinMode(DIR, OUTPUT);
+    pinMode(PWM, OUTPUT);
+    pinMode(GENEVA_ENCODER_A_PIN, INPUT_PULLDOWN);
+    pinMode(GENEVA_ENCODER_B_PIN, INPUT_PULLDOWN);
+
+    attachInterrupt(digitalPinToInterrupt(GENEVA_ENCODER_A_PIN), pulseEncoderA, RISING);
+    attachInterrupt(digitalPinToInterrupt(GENEVA_ENCODER_B_PIN), pulseEncoderB, RISING);
+
+    timer = timerBegin(GENEVA_TIMER, 80, true);
+
+    timerAttachInterrupt(timer, &onTimer, true);
+    // call onTimer every 0.1 seconds
+    timerAlarmWrite(timer, 100000, true);
+    timerAlarmEnable(timer);
+}
+
+void IRAM_ATTR Geneva::pulseEncoderA()
+{
+    count_a = count_a + dir;
 }
 
-float Geneva::getCurrentAngle()
+void IRAM_ATTR Geneva::pulseEncoderB()
 {
-    // TODO(#2458): convert from EncoderValue to angle
-    return 0;
+    count_b = count_b + dir;
 }
 
-void Geneva::performWhenDone(void (*isr)(void))
+void IRAM_ATTR Geneva::onTimer()
 {
-    timerRestart(timer);
+    // stop motor when it stalls aka in side slots
+    // or if its centering it should be a fixed amount
+    if (((currentSlot == TbotsProto_Geneva_Slot_LEFT ||
+          currentSlot == TbotsProto_Geneva_Slot_RIGHT) &&
+         ((count_a != 0 && prev_count_a == count_a) ||
+          (count_b != 0 && prev_count_b == count_b))) ||
```

Comment:
here is a suggestion to make this more readable + safe,

~~~
// If the encoder is not incrementing, we know the motor is stalled
bool geneva_motor_stalled = ((count_a != 0 && prev_count_a == count_a) || (count_b != 0 && prev_count_b == count_b));

// If the motor has stalled, we don't want to keep applying PWM even if we are not at the end
if (geneva_motor_stalled)
{
     digitalWrite(PWM, LOW);
}

if (currentSlot == TbotsProto_Geneva_Slot_LEFT || currentSlot == TbotsProto_Geneva_Slot_RIGHT ...)
{
     count_a = 0;
     count_b = 0;
     callbacks...
}

~~~